### PR TITLE
docs: translate keeping-components-pure

### PR DESCRIPTION
**Pull request by viboloveyou12**
Hi, Please let me know if any problem I need to fix.
Thank you :)

### CODE REVIEW

**Comment by neighborhood999** [comment]
> Hi, Please let me know if any problem I need to fix. Thank you :)

Hi @viboloveyou12,

First, thank you for your contribution. I found @rccj claimed this document to translate😅, but it's okay, and don't worry. If you want to contribute other documents don't forget to leave a comment on #260, I will update the list to avoid this situation.

**Comment by viboloveyou12** [comment]
Oh, I'm so sorry that I didn't notice the issue!
Thank you for reminding me of this, I'll definitely check it and leave a comment next time.🙏

**Comment by rccj** [comment]
@neighborhood999 hi, I just noticed this mention now. Please remove me so that @viboloveyou12 can complete this contribution without any concerns. I'm sorry I couldn't complete this task.

File: src/content/learn/keeping-components-pure.md [hunk 1]
```diff
@@ -185,59 +186,59 @@ export default function TeaGathering() {
...
+#### 為什麼 React 在意存粹性？ {/*why-does-react-care-about-purity*/}
 
-Writing pure functions takes some habit and discipline. But it also unlocks marvelous opportunities:
+撰寫純函數需要一些習慣與紀律。但純函數也解鎖了一些絕佳的功能：
 
-* Your components could run in a different environment—for example, on the server! Since they return the same result for the same inputs, one component can serve many user requests.
-* You can improve performance by [skipping rendering](/reference/react/memo) components whose inputs have not changed. This is safe because pure functions always return the same results, so they are safe to cache.
-* If some data changes in the middle of rendering a deep component tree, React can restart rendering without wasting time to finish the outdated render. Purity makes it safe to stop calculating at any time.
+* 你的 components 可以在不同環境上運行 - 例如，在伺服器上！由於它們對相同輸出會有相同結果，因此一個 component 可以滿足許多用戶請求。
+* 你可以透過 [跳過渲染](/reference/react/memo) 那些 input 沒有更新的 components 來提昇性能。這是安全的，因為純函數永遠都會回傳相同的結果，所以可以安全地快取它們。
```
Comment (neighborhood999) [inline src/content/learn/keeping-components-pure.md:210]: we reserve skipping rendering here
```suggestion
* 你可以透過 [skipping rendering](/reference/react/memo) 那些 input 沒有更新的 components 來提升效能。這是安全的，因為純函式永遠都會回傳相同的結果，所以可以安全地 cache 它們。
```

File: src/content/learn/keeping-components-pure.md [hunk 2]
```diff
@@ -185,59 +186,59 @@ export default function TeaGathering() {
...
 
-* A component must be pure, meaning:
-  * **It minds its own business.** It should not change any objects or variables that existed before rendering.
-  * **Same inputs, same output.** Given the same inputs, a component should always return the same JSX. 
-* Rendering can happen at any time, so components should not depend on each others' rendering sequence.
-* You should not mutate any of the inputs that your components use for rendering. That includes props, state, and context. To update the screen, ["set" state](/learn/state-a-components-memory) instead of mutating preexisting objects.
-* Strive to express your component's logic in the JSX you return. When you need to "change things", you'll usually want to do it in an event handler. As a last resort, you can `useEffect`.
-* Writing pure functions takes a bit of practice, but it unlocks the power of React's paradigm.
+* 一個 component 是 pure 的，這意味著：
+  * **不多管閒事** 這個函數不會修改任何在它被呼叫之前就已經存在的物件或變數。
```
Comment (neighborhood999) [inline src/content/learn/keeping-components-pure.md:220]:
```suggestion
  * **只關心自己的事務。**這個函式不會修改任何在他被呼叫之前就已經存在的 object 或變數。
```

File: src/content/learn/keeping-components-pure.md [hunk 3]
```diff
@@ -185,59 +186,59 @@ export default function TeaGathering() {
...
 
 
   
 <Challenges>
 
-#### Fix a broken clock {/*fix-a-broken-clock*/}
+#### 修正一個換掉的時鐘 {/*fix-a-broken-clock*/}
 
-This component tries to set the `<h1>`'s CSS class to `"night"` during the time from midnight to six hours in the morning, and `"day"` at all other times. However, it doesn't work. Can you fix this component?
+這個 component 希望在午夜至上午 6 點期間將 `<h1>` 的 CSS class 設置為 `"night"`，並在所有其他時段都設成 `"day"`。但是它沒辦法運作。你能修復這個 component 嗎？
```
Comment (neighborhood999) [inline src/content/learn/keeping-components-pure.md:235]:
```suggestion
這個 component 希望在午夜至上午 6 點期間將 `<h1>` 的 CSS class 設定為 `"night"`，並在所有其他時段都設成 `"day"`。但是它沒辦法運作。你能修正這個 component 嗎？
```

**Review by neighborhood999 — COMMENTED** [review]
Hi @viboloveyou12 ,

Thank you for your hard work 🙌, I gave you some suggestions and you can take a look.
Let me know if you have any questions about my suggestions; most translation looks good to me; For some terms we can reserve, you can check [Glossary](https://github.com/reactjs/zh-hant.react.dev/wiki/Glossary),.

File: src/content/learn/keeping-components-pure.md [hunk 4]
```diff
@@ -141,31 +142,31 @@ export default function TeaSet() {
...
+請注意在原本的範例，它顯示了「Guest #2」、 「Guest #4」以及「Guest #6」，而不是「Guest #1」、「Guest #2」及「Guest #3」。原本的函數是 impure 的，所以呼叫兩次後就破壞了它。但在修正後的 pure 版本中，即使每次呼叫了兩次函數還是能夠正常運作。 **純函數只進行運算，因此呼叫兩次後也不會改變任何事** -- 就像是呼叫 `double(2)` 兩次也不會改變它的回傳值，求解 <Math><MathI>y</MathI> = 2<MathI>x</MathI></Math> 兩次不會改變 <MathI>y</MathI> 的值。相同的輸入永遠會有相同的輸出。
 
-Strict Mode has no effect in production, so it won't slow down the app for your users. To opt into Strict Mode, you can wrap your root component into `<React.StrictMode>`. Some frameworks do this by default.
+嚴格模式不會影響正式環境，因此它不會拖慢用戶的應用程式速度。如需選擇嚴格模式，你可以將你的 root component 包裝到 `<React.StrictMode>`。有些框架預設會這麼做。
 
 </DeepDive>
 
-### Local mutation: Your component's little secret {/*local-mutation-your-components-little-secret*/}
+### 變異本地化：你的 component 的小秘密 {/*local-mutation-your-components-little-secret*/}
```
[the remark this answers is not in the document]
Comment (viboloveyou12) [reply src/content/learn/keeping-components-pure.md:165]: Yes, I agree. We should just keep the original here. Thanks!

File: src/content/learn/keeping-components-pure.md [hunk 5]
```diff
@@ -141,31 +142,31 @@ export default function TeaSet() {
...
 <DeepDive>
 
-#### Detecting impure calculations with StrictMode {/*detecting-impure-calculations-with-strict-mode*/}
+#### 使用嚴格模式檢查 impure 的計算 {/*detecting-impure-calculations-with-strict-mode*/}
 
-Although you might not have used them all yet, in React there are three kinds of inputs that you can read while rendering: [props](/learn/passing-props-to-a-component), [state](/learn/state-a-components-memory), and [context.](/learn/passing-data-deeply-with-context) You should always treat these inputs as read-only.
+僅管你可能還沒有全部使用過它們，但在 React 中你可以在渲染時讀取三種輸入：[props](/learn/passing-props-to-a-component)、[state](/learn/state-a-components-memory) 以及 [context](/learn/passing-data-deeply-with-context)。你應該永遠將這些輸入視為 read-only 。
 
-When you want to *change* something in response to user input, you should [set state](/learn/state-a-components-memory) instead of writing to a variable. You should never change preexisting variables or objects while your component is rendering.
+當你想要 *改變* 某些以用戶輸入為響應的內容時，你應該要 [set state](/learn/state-a-components-memory) 而非直接更改變數。你永遠都不該在 component 渲染過程中改變已存在的變數或物件。
```
[the remark this answers is not in the document]
Comment (viboloveyou12) [reply src/content/learn/keeping-components-pure.md:155]: I'm not sure if the "為" word was mistyped in the sentence 「當你想要*改變*某些內容來回應使用者輸入為時」, so I will modify it without this word. If there is anything I misunderstand please let me know!

**Comment by viboloveyou12** [comment]
Hi @neighborhood999 ,
Thanks for your review and suggestions.
I've modified a new version, if there is anything that can be improved please let me know.

Thank you:)

File: src/content/learn/keeping-components-pure.md [hunk 6]
```diff
@@ -185,59 +186,59 @@ export default function TeaGathering() {
...
 
-Writing pure functions takes some habit and discipline. But it also unlocks marvelous opportunities:
+撰寫純函式需要一些習慣與紀律。但純函式也解鎖了一些絕佳的功能：
 
-* Your components could run in a different environment—for example, on the server! Since they return the same result for the same inputs, one component can serve many user requests.
-* You can improve performance by [skipping rendering](/reference/react/memo) components whose inputs have not changed. This is safe because pure functions always return the same results, so they are safe to cache.
-* If some data changes in the middle of rendering a deep component tree, React can restart rendering without wasting time to finish the outdated render. Purity makes it safe to stop calculating at any time.
+* 你的 component 可以在不同環境上運行 - 例如，在伺服器上！由於它們對相同輸出會有相同結果，因此一個 component 可以滿足許多用戶請求。
+* 你可以透過 [skipping rendering](/reference/react/memo) 那些 input 沒有更新的 component 來提升效能。這是安全的，因為純函式永遠都會回傳相同的結果，所以可以安全地 cache 它們。
+* 如果在 render 深層元件樹 (deep component tree) 的過程中某些資料發生變化，React 可以重新進行 render、而不浪費時間完成過時的 render。 Purity 可以讓它更安全地隨時停止計算。
```
Comment (neighborhood999) [inline src/content/learn/keeping-components-pure.md:211]:
```suggestion
* 如果在 render 一個 deep component tree 的過程中某些資料發生變化，React 可以重新進行 render、而不浪費時間完成過時的 render。Purity 可以讓它更安全地隨時停止計算。
```

File: src/content/learn/keeping-components-pure.md [hunk 7]
```diff
@@ -185,59 +186,59 @@ export default function TeaGathering() {
...
-#### Why does React care about purity? {/*why-does-react-care-about-purity*/}
+#### 為什麼 React 在意 purity？ {/*why-does-react-care-about-purity*/}
 
-Writing pure functions takes some habit and discipline. But it also unlocks marvelous opportunities:
+撰寫純函式需要一些習慣與紀律。但純函式也解鎖了一些絕佳的功能：
 
-* Your components could run in a different environment—for example, on the server! Since they return the same result for the same inputs, one component can serve many user requests.
-* You can improve performance by [skipping rendering](/reference/react/memo) components whose inputs have not changed. This is safe because pure functions always return the same results, so they are safe to cache.
-* If some data changes in the middle of rendering a deep component tree, React can restart rendering without wasting time to finish the outdated render. Purity makes it safe to stop calculating at any time.
+* 你的 component 可以在不同環境上運行 - 例如，在伺服器上！由於它們對相同輸出會有相同結果，因此一個 component 可以滿足許多用戶請求。
```
Comment (neighborhood999) [inline src/content/learn/keeping-components-pure.md:209]:
```suggestion
* 你的 component 可以在不同環境上運行 - 例如，在伺服器上！由於它們對相同輸出會有相同結果，因此一個 component 可以滿足許多使用者請求。
```

File: src/content/learn/keeping-components-pure.md [hunk 8]
```diff
@@ -185,59 +186,59 @@ export default function TeaGathering() {
...
+* 你可以透過 [skipping rendering](/reference/react/memo) 那些 input 沒有更新的 component 來提升效能。這是安全的，因為純函式永遠都會回傳相同的結果，所以可以安全地 cache 它們。
+* 如果在 render 深層元件樹 (deep component tree) 的過程中某些資料發生變化，React 可以重新進行 render、而不浪費時間完成過時的 render。 Purity 可以讓它更安全地隨時停止計算。
 
-Every new React feature we're building takes advantage of purity. From data fetching to animations to performance, keeping components pure unlocks the power of the React paradigm.
+所有我們正在建立的 React 新功能都利用了 purity 的優點。從獲取資料到動畫再到性能，保持 component 的 purity 能夠解鎖 React 典範的力量。
```
Comment (neighborhood999) [inline src/content/learn/keeping-components-pure.md:213]:
```suggestion
所有我們正在建立的 React 新功能都利用了 purity 的優點。從獲取資料到動畫再到效能，保持 component 的 purity 能夠解鎖 React 典範的力量。
```

File: src/content/learn/keeping-components-pure.md [hunk 9]
```diff
@@ -185,59 +186,59 @@ export default function TeaGathering() {
...
-* Rendering can happen at any time, so components should not depend on each others' rendering sequence.
-* You should not mutate any of the inputs that your components use for rendering. That includes props, state, and context. To update the screen, ["set" state](/learn/state-a-components-memory) instead of mutating preexisting objects.
-* Strive to express your component's logic in the JSX you return. When you need to "change things", you'll usually want to do it in an event handler. As a last resort, you can `useEffect`.
-* Writing pure functions takes a bit of practice, but it unlocks the power of React's paradigm.
+* 一個 component 是 pure 的，這意味著：
+  * **只關心自己的事務。** 這個函式不會修改任何在他被呼叫之前就已經存在的 object 或變數。
+  * **一樣的輸入，一樣的輸出** 只要我們輸入相同的參數，這個函式總是回傳一個相同的輸出。
+* Rendering 可能會在任何時間發生，因此 component 不該依賴於彼此的 rendering 順序。
+* 你不該改變任何你的 component 用來 render 的輸入。這包含 props，state，以及 context。要更新畫面的話，請 ["set" state](/learn/state-a-components-memory) 而不是直接修改預先存在的 object。
+* 盡量在返回的 JSX 中表達你的 component 邏輯。當你需要「更改內容」時，你會希望在 event handler 中處理。或是作為最後的手段，你可以使用 `useEffect`。
```
Comment (neighborhood999) [inline src/content/learn/keeping-components-pure.md:224]:
```suggestion
* 盡量在回傳的 JSX 中表達你的 component 邏輯。當你需要「更改內容」時，你會希望在 event handler 中處理。或是作為最後的手段，你可以使用 `useEffect`。
```

File: src/content/learn/keeping-components-pure.md [hunk 10]
```diff
@@ -362,19 +363,19 @@ body > * {
 
 </Sandpack>
 
-In this example, the side effect (modifying the DOM) was not necessary at all. You only needed to return JSX.
+在這個範例中，side effect（修改 DOM ）根本完全沒有必要。你只需要返回 JSX 即可。
```
Comment (neighborhood999) [inline src/content/learn/keeping-components-pure.md:366]:
```suggestion
在這個範例中，side effect（修改 DOM ）根本完全沒有必要。你只需要回傳 JSX 即可。
```

File: src/content/learn/keeping-components-pure.md [hunk 11]
```diff
@@ -362,19 +363,19 @@ body > * {
...
 
 </Solution>
 
-#### Fix a broken profile {/*fix-a-broken-profile*/}
+#### 修正一個壞掉的 Profile {/*fix-a-broken-profile*/}
 
-Two `Profile` components are rendered side by side with different data. Press "Collapse" on the first profile, and then "Expand" it. You'll notice that both profiles now show the same person. This is a bug.
+有兩個 `Profile` component 使用不同資料並排地呈現。在第一個 Profile 中點選 "Collapse"，接著 "Expand"。你會發現這時兩個 profiles 顯示的是同一個人。這是一個 bug。
```
Comment (neighborhood999) [inline src/content/learn/keeping-components-pure.md:372]:
```suggestion
有兩個 `Profile` component 使用不同資料並排地呈現。在第一個 Profile 中點選「Collapse」，接著 「Expand」。你會發現這時兩個 profiles 顯示的是同一個人。這是一個 bug。
```

File: src/content/learn/keeping-components-pure.md [hunk 12]
```diff
@@ -571,15 +572,15 @@ h1 { margin: 5px; font-size: 18px; }
 
 </Sandpack>
 
-Remember that React does not guarantee that component functions will execute in any particular order, so you can't communicate between them by setting variables. All communication must happen through props.
+請記住， React 不保證 component 函式會以任何特定順序執行，所以你無法透過設置變數在它們之間溝通。所有的溝通都必須透過 props 進行。
```
Comment (neighborhood999) [inline src/content/learn/keeping-components-pure.md:575]:
```suggestion
請記住， React 不保證 component 函式會以任何特定順序執行，所以你無法透過設定變數在它們之間溝通。所有的溝通都必須透過 props 進行。
```

File: src/content/learn/keeping-components-pure.md [hunk 13]
```diff
@@ -571,15 +572,15 @@ h1 { margin: 5px; font-size: 18px; }
...
 
 </Solution>
 
-#### Fix a broken story tray {/*fix-a-broken-story-tray*/}
+#### 修正一個壞掉的故事列表 {/*fix-a-broken-story-tray*/}
 
-The CEO of your company is asking you to add "stories" to your online clock app, and you can't say no. You've written a `StoryTray` component that accepts a list of `stories`, followed by a "Create Story" placeholder.
+你公司的 CEO 要求你將「多個故事」加入線上時鐘應用程式中，而你不能拒絕。你已經編寫了一個 `StoryTray` component，它接受「故事」列表，並會在後面接上一個 "Create Story" 的圖片框。
```
Comment (neighborhood999) [inline src/content/learn/keeping-components-pure.md:581]:
```suggestion
你公司的 CEO 要求你將「多個故事」加入線上時鐘應用程式中，而你不能拒絕。你已經編寫了一個 `StoryTray` component，它接受「故事」列表，並會在後面接上一個「Create Story」的圖片框。
```

File: src/content/learn/keeping-components-pure.md [hunk 14]
```diff
@@ -571,15 +572,15 @@ h1 { margin: 5px; font-size: 18px; }
...
 
-You implemented the "Create Story" placeholder by pushing one more fake story at the end of the `stories` array that you receive as a prop. But for some reason, "Create Story" appears more than once. Fix the issue.
+你透過在作為 props 的 `stories` array 後面加上一筆假資料來實作 "Create Story" 的圖片框。但由於某種原因，"Create Story" 出現了不只一次，請解決這個問題。
```
Comment (neighborhood999) [inline src/content/learn/keeping-components-pure.md:583]:
```suggestion
你透過在作為 props 的 `stories` array 後面加上一筆假資料來實作「Create Story」的圖片框。但由於某種原因，「Create Story」出現了不只一次，請解決這個問題。
```

File: src/content/learn/keeping-components-pure.md [hunk 15]
```diff
@@ -675,11 +676,11 @@ li {
...
 <Solution>
 
-Notice how whenever the clock updates, "Create Story" is added *twice*. This serves as a hint that we have a mutation during rendering--Strict Mode calls components twice to make these issues more noticeable.
+請注意，每當時鐘更新時，"Create Story" 就會被增加 *兩次* 。這暗示我們在 render 過程中發生了一個變異 -- Strict Mode 調用了兩次 component 使得這個問題更明顯。
 
-`StoryTray` function is not pure. By calling `push` on the received `stories` array (a prop!), it is mutating an object that was created *before* `StoryTray` started rendering. This makes it buggy and very difficult to predict.
+`StoryTray` 函式不是 pure 的。透過在接收到的 `stories` array（一個 prop ）呼叫 `push` ，它會改變在 `StoryTray` render 前就建立的 object 。這使得它變得充滿錯誤並且難以預測。
 
-The simplest fix is to not touch the array at all, and render "Create Story" separately:
+最簡單的修正作法是完全不要修改 array，只單獨渲染 "Create Story"：
```
Comment (neighborhood999) [inline src/content/learn/keeping-components-pure.md:683]:
```suggestion
最簡單的修正作法是完全不要修改 array，只單獨 render「Create Story」：
```

File: src/content/learn/keeping-components-pure.md [hunk 16]
```diff
@@ -763,7 +764,7 @@ li {
 
 </Sandpack>
 
-Alternatively, you could create a _new_ array (by copying the existing one) before you push an item into it:
+或者，你可以在你推入新的項目前創立一個 _新的_ array（透過複製現有 array ）：
```
Comment (neighborhood999) [inline src/content/learn/keeping-components-pure.md:767]:
```suggestion
或者，你可以在你推入新的項目前建立一個_新的_ array（透過複製現有 array ）：
```

File: src/content/learn/keeping-components-pure.md [hunk 17]
```diff
@@ -855,9 +856,9 @@ li {
 
 </Sandpack>
 
-This keeps your mutation local and your rendering function pure. However, you still need to be careful: for example, if you tried to change any of the array's existing items, you'd have to clone those items too.
+這使得變異能夠保持本地化、並且保持渲染函式純粹。但是你仍然需要小心：舉例來說，如果你嘗試改變 array 中的任何現有項目，你也必須先複製這些項目。
```
Comment (neighborhood999) [inline src/content/learn/keeping-components-pure.md:859]:
```suggestion
這使得 mutation 能夠保持在 local 並且保持渲染函式純粹。但是你仍然需要小心：例如，如果你嘗試改變 array 中的任何現有項目，你也必須先複製這些項目。
```

**Review by neighborhood999 — COMMENTED** [review]
Thank @viboloveyou12 for helping, most look great to me.

**Review by neighborhood999 — APPROVED** [review]

**Pull request merged**